ADBDEV-3913: Refactoring. Declare methods as static if they work with static member variables only

RekGRpth opened the pull request:
Refactoring. Declare methods as static if they work with static member variables only

- I made the CCacheFactory::Shutdown method static because it only uses static variables.
- I made the CFSimulator::Shutdown method static, which was previously called only on the static object m_fsim.
- I made the CMemoryPoolManager::Cleanup method static, which was previously called only on the static object m_memory_pool_mgr.
- I made the CMemoryPoolManager::CreateMemoryPool method static, which was previously called only on the static object m_memory_pool_mgr.
- I made the CMemoryPoolManager::Destroy method static, which was previously called only on the static object m_memory_pool_mgr.
- I made the CMemoryPoolManager::GetGlobalMemoryPool method static, which was previously called only on the static object m_memory_pool_mgr.
- I made the CMemoryPoolManager::Shutdown method static, which was previously called only on the static object m_memory_pool_mgr.
- I made the CMessageRepository::Shutdown method static, which was previously called on the static object m_repository.
- I made the CWorkerPoolManager::Shutdown method static because it uses only static variables.
- I made the CXformFactory::Shutdown method static because it uses only static variables.

I added assertions to make sure the pointers being dereferenced are not null.

andr-sokolov commented on src/backend/gporca/libgpos/include/gpos/memory/CCacheFactory.h:
```diff
@@ -60,15 +60,13 @@ class CCacheFactory
 	// private dtor
 	~CCacheFactory()
```
Destructor has become empty. Let's delete it

RekGRpth replied:
done

andr-sokolov commented on src/backend/gporca/libgpos/include/gpos/memory/CMemoryPoolManager.h:
```diff
@@ -130,27 +130,27 @@ class CMemoryPoolManager
 	GetGlobalMemoryPool()
 	{
-		return m_global_memory_pool;
+		return m_memory_pool_mgr->m_global_memory_pool;
```
Describe please in PR description why the return value has been changed

RekGRpth replied:
> Describe please in PR description why the return value has been changed

The return value has NOT been changed!

andr-sokolov replied:
ok. describe why m_global_memory_pool = m_memory_pool_mgr->m_global_memory_pool

RekGRpth replied:
> why m_global_memory_pool = m_memory_pool_mgr->m_global_memory_pool

Isn't it obvious?

andr-sokolov commented on src/backend/gporca/libgpos/include/gpos/task/CWorkerPoolManager.h:
```diff
@@ -131,8 +131,6 @@ class CWorkerPoolManager
 	// dtor
 	~CWorkerPoolManager()
```
Describe please in PR description why it is safe to remove the assert. I suggest to delete this destructor too.

RekGRpth replied:
> I suggest to delete this destructor too.

done

RekGRpth replied:
> Describe please in PR description why it is safe to remove the assert.

I restored assertion I deleted.

andr-sokolov commented on src/backend/gporca/libgpos/src/error/CFSimulator.cpp:
```diff
@@ -158,11 +158,11 @@ CFSimulator::Init()
-	CMemoryPool *mp = m_mp;
-	GPOS_DELETE(CFSimulator::m_fsim);
-	CFSimulator::m_fsim = NULL;
+	CMemoryPool *mp = m_fsim->m_mp;
```
Describe please in PR description why m_mp was replaced with m_fsim->m_mp

RekGRpth replied:
m_mp was replaced with m_fsim->m_mp because m_mp is a field of an instance of the CFSimulator class, which is precisely m_fsim.

andr-sokolov commented on src/backend/gporca/libgpos/src/memory/CMemoryPoolManager.cpp:
```diff
@@ -82,14 +82,14 @@ CMemoryPoolManager::Init()
 CMemoryPoolManager::CreateMemoryPool()
 {
-	CMemoryPool *mp = NewMemoryPool();
+	CMemoryPool *mp = m_memory_pool_mgr->NewMemoryPool();
```
Describe please in PR description why m_memory_pool_mgr appears here and below

RekGRpth replied:
> why m_memory_pool_mgr appears here and below

Because I changed this method to static.

andr-sokolov commented on src/backend/gporca/libgpos/include/gpos/memory/CCacheFactory.h:
```diff
@@ -57,18 +57,11 @@ class CCacheFactory
 
 public:
-	// private dtor
-	~CCacheFactory()
```
Describe why it is safe to delete this destructor

RekGRpth replied:
> why it is safe to delete this destructor

As far as I understand, if the destructor is not specified, then the compiler will create it itself to free the memory of the members of the class instance.
Therefore, after all, an explicit empty destructor is needed.

RekGRpth commented:
@andr-sokolov, I restored both assertions I deleted.

RekGRpth committed:
Refactoring. Declare methods as static if they work with static
member variables only

- I made the CCacheFactory::Shutdown method static,
because it only uses static variables.

- I made the CFSimulator::Shutdown method static,
which was previously called only on the static object m_fsim.

- I made the CMemoryPoolManager::Cleanup method static,
which was previously called only on the static object m_memory_pool_mgr.

- I made the CMemoryPoolManager::CreateMemoryPool method static,
which was previously called only on the static object m_memory_pool_mgr.

- I made the CMemoryPoolManager::Destroy method static,
which was previously called only on the static object m_memory_pool_mgr.

- I made the CMemoryPoolManager::GetGlobalMemoryPool method static,
which was previously called only on the static object m_memory_pool_mgr.

- I made the CMemoryPoolManager::Shutdown method static,
which was previously called only on the static object m_memory_pool_mgr.

- I made the CMessageRepository::Shutdown method static,
which was previously called on the static object m_repository.

- I made the CWorkerPoolManager::Shutdown method static,
because it uses only static variables.

- I made the CXformFactory::Shutdown method static,
because it uses only static variables.

I added assertions to make sure the pointers being dereferenced are not null.

HustonMmmavr commented on src/backend/gporca/libgpos/include/gpos/memory/CCacheFactory.h:
```diff
@@ -84,10 +84,10 @@ class CCacheFactory
 				typename CCache<T, K>::HashFuncPtr hash_func,
 				typename CCache<T, K>::EqualFuncPtr equal_func)
 	{
-		GPOS_ASSERT(NULL != GetFactory() &&
```
Seems, after these changes (ar CCacheFactory.h/cpp) the GetFactory method is useless. Maybe drop it?

RekGRpth replied:
> after this changes the GetFactory method is useless. Maybe drop it?

No, it's needed for the next [patch](https://github.com/arenadata/gpdb/pull/525/files#diff-2bffde632791e7b2be2e152860e63cedcdf6720548caa6db7cee6dfa8771710cR174).

HustonMmmavr commented on src/backend/gporca/libgpos/src/memory/CCacheFactory.cpp:
```diff
@@ -60,25 +60,23 @@ CCacheFactory::Pmp() const
+		CMemoryPoolManager::Destroy(mp);
 
-		CCacheFactory::m_factory = NULL;
+		m_factory = NULL;
```
Does this needed, seems after throwed exceprtion - there would be null, am I wrong?
This comment (with a related question) has a better description  https://github.com/arenadata/gpdb/pull/555#discussion_r1257528940

HustonMmmavr replied:
This comment (with a related question) has a better description  https://github.com/arenadata/gpdb/pull/555#discussion_r1257528940

HustonMmmavr commented on src/backend/gporca/libgpopt/include/gpopt/xforms/CXformFactory.h:
```diff
@@ -114,7 +114,7 @@ class CXformFactory
 
 	// destroy global factory instance
-	void Shutdown();
+	static void Shutdown();
```
Maybe it's worth to convert and other methods of this class like `Pxf`, `PxfsExploration`, `PxfsImplementation` into static (seems their usage may be static) and remove usage of global accessor of static `m_pxff` variable - `Pxff()` method. And for the next patch ADBDEV-3674 instead of check `NULL == CXFormFactory::PXff()` implement method `IsInited`, like
```c
static BOOL Inited()
{
     return NULL != m_pxff
}
```
Yes, It'll increase the diff, but seems it would be more brief and readable, how do you think?

RekGRpth replied:
I changed to static only those methods that are needed for use in the next [patch](https://github.com/arenadata/gpdb/pull/525).

HustonMmmavr commented on src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp:
```diff
@@ -371,7 +370,7 @@ CXformFactory::Init()
 		// destroy memory pool if global instance was not created
-		CMemoryPoolManager::GetMemoryPoolMgr()->Destroy(mp);
+		CMemoryPoolManager::Destroy(mp);
 		m_pxff = NULL;
```
Do we really need to set the static variable to NULL (and at other classes, related to the patch, used the same approach, does it really needed?)? In case of raised exception the value of the static variable, seems  doesn't change. Moreover, this variable is inited globally with NULL the Shutdown method may set it to NULL. So this instruction (and similar ones at other classes, related to the patch) seems redundant

HustonMmmavr replied:
This command seems useless, but let it be

HustonMmmavr commented on src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp:
```diff
@@ -405,18 +404,18 @@ CXformFactory::Init()
 
 	GPOS_ASSERT(NULL != pxff && "Xform factory has not been initialized");
 
 	CMemoryPool *mp = pxff->m_mp;
```
Does the order of the commands, executed inside Shutdown method (not only in this class, but at whole classes changed in this patch) is significant? In case of raised exception on the release build, the static `m_pxff` would be set to NULL, but other classes has a different order (for ex. [here](https://github.com/arenadata/gpdb/blob/5681fe94a83080e60d8b361bd9b6e8d6bbe90969/src/backend/gporca/libgpos/src/common/CDebugCounter.cpp#L79), or the CMessageRepository), and the static variable may be not cleaned up. For me it's a bit confusing, may be we shouldn't do the same code at the shutdown methods, but we should use the single order of cleanup?

RekGRpth replied:
> Does the order of the commands, executed inside Shutdown method (not only in this class, but at whole classes changed in this patch) is significant?

Yes, because some destructors have corresponding asserts.

HustonMmmavr replied:
I understood, other classes which doesn't have an assertion like `GPOS_ASSERT(global_instalnce == NULL)` inside descturcor may not nulls the variable as a first command, so them may call the possble  method (which may throw exception) before nulling, and seems it's a valid case: Shutdown method may throw an exception before it nulls the global_instance, but the desctructor won't fail on assertion (on debug build) for example [CFSImulator::Shutdown](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/src/error/CFSimulator.cpp#L158-L166) may throw an exception on GPOS_DEBUG, but it's [destructor](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/include/gpos/error/CFSimulator.h#L165-L167) is empty

Stolb27 commented:
I think, it would be nice to provide motivation of refactoring in commit and PR description.

andr-sokolov commented:
> The main goal was, as stated in the description, to convert to static methods that work only with static variables.

@RekGRpth , The goal is to make the [PR](https://github.com/arenadata/gpdb/pull/525) more readable and avoid errors in the future. Without this PR we need to call a non-static method for NULL pointer. The method use static members only and the call is ok now, but someday the method may be changed and it will use non-static members and it will cause to segfault. What is the method I write about and where is it called?

RekGRpth commented:
> someday the method may be changed and it will use non-static members and it will cause to segfault

There will be no segfault, because this code won't even compile!

RekGRpth commented:
> Without this PR we need to call a non-static method for NULL pointer.

No! Previously, the pointer to the object, on which the method was then called, was not checked for null.

RekGRpth commented:
> What is the method I write about and where is it called?

CMemoryPoolManager::GetMemoryPoolMgr()?
